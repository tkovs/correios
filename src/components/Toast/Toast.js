import React from 'react'
import PropTypes from 'prop-types'
import { Snackbar } from 'react-native-paper'

function Toast({ duration, message, onDismiss, visible }) {
  return (
    <Snackbar visible={visible} onDismiss={onDismiss} duration={duration}>
      {message}
    </Snackbar>
  )
}

Toast.propTypes = {
  duration: PropTypes.number.isRequired,
  message: PropTypes.string.isRequired,
  onDismiss: PropTypes.func.isRequired,
  visible: PropTypes.bool.isRequired,
}

export default Toast
