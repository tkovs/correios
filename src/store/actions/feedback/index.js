import * as types from './types'

export const clearFeedback = () => ({
  type: types.CLEAR_FEEDBACK,
})

export const addFeedback = message => ({
  type: types.ADD_FEEDBACK,
  payload: {
    message,
  },
})
