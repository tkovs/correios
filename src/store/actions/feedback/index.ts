import {
  ADD_FEEDBACK,
  AddFeedback,
  CLEAR_FEEDBACK,
  ClearFeedback,
  REJECT_CLEAR_FEEDBACK,
  RejectClearFeedback,
} from './types'

export const clearFeedback = (): ClearFeedback => ({
  type: CLEAR_FEEDBACK,
})

export const rejectClearFeedback = (): RejectClearFeedback => ({
  type: REJECT_CLEAR_FEEDBACK,
})

export const addFeedback = (message: string): AddFeedback => ({
  type: ADD_FEEDBACK,
  payload: {
    message,
  },
})
