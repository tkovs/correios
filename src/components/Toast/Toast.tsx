import React, { FC } from 'react'
import { Snackbar } from 'react-native-paper'
import { FeedbackActionTypes } from '../../store/actions/feedback/types'

interface ToastProps {
  duration: number
  message: string
  visible: boolean
  onDismiss: () => FeedbackActionTypes
}

type Props = ToastProps

const Toast: FC<Props> = ({ duration, message, onDismiss, visible }: Props) => (
  <Snackbar
    action={{
      label: 'Ok!',
      onPress: onDismiss,
    }}
    visible={visible}
    onDismiss={onDismiss}
    duration={duration}
    accessibilityStates={['disabled']}
  >
    {message}
  </Snackbar>
)

export default Toast
