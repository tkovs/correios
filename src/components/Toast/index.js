import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import Toast from './Toast'
import { clearFeedback } from '../../store/actions/feedback'
import {
  messageSelector,
  visibleSelector,
} from '../../store/selectors/feedback'

function ToastContainer({ hideToast, message, visible }) {
  return (
    <Toast
      message={message}
      visible={visible}
      onDismiss={hideToast}
      duration={3000}
    />
  )
}

ToastContainer.propTypes = {
  hideToast: PropTypes.func.isRequired,
  message: PropTypes.string.isRequired,
  visible: PropTypes.bool.isRequired,
}

const mapStateToProps = state => ({
  message: messageSelector(state),
  visible: visibleSelector(state),
})

const mapDispatchToProps = dispatch => ({
  hideToast: () => dispatch(clearFeedback()),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ToastContainer)
