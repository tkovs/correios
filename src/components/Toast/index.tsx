import React, { FC } from 'react'
import { connect, MapStateToProps, MapDispatchToProps } from 'react-redux'

import Toast from './Toast'
import { clearFeedback } from '../../store/actions/feedback'
import {
  messageSelector,
  visibleSelector,
} from '../../store/selectors/feedback'
import { FeedbackActionTypes } from '../../store/actions/feedback/types'

interface StateProps {
  message: string
  visible: boolean
}

interface DispatchProps {
  hideToast: () => FeedbackActionTypes
}

type Props = DispatchProps & StateProps

const ToastContainer: FC<Props> = ({ hideToast, message, visible }: Props) => {
  return (
    <Toast
      message={message}
      visible={visible}
      onDismiss={hideToast}
      duration={3000}
    />
  )
}

const mapStateToProps: MapStateToProps<StateProps, {}, State> = (
  state: State
) => ({
  message: messageSelector(state),
  visible: visibleSelector(state),
})

const mapDispatchToProps: MapDispatchToProps<DispatchProps, {}> = (
  dispatch
) => ({
  hideToast: () => dispatch(clearFeedback()),
})

export default connect(mapStateToProps, mapDispatchToProps)(ToastContainer)
