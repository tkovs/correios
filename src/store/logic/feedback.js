import { createLogic } from 'redux-logic'

import { CLEAR_FEEDBACK } from '../actions/feedback/types'

const clearFeedbackLogic = createLogic({
  type: CLEAR_FEEDBACK,
  validate: ({ getState, action }, allow, reject) => {
    const { message, visible } = getState().feedback

    if (message !== '' || visible === true) {
      allow(action)
    } else {
      reject()
    }
  },
})

export default [clearFeedbackLogic]
