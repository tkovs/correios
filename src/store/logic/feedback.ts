import { createLogic } from 'redux-logic'

import { CLEAR_FEEDBACK } from '../actions/feedback/types'
import { rejectClearFeedback } from '../actions/feedback'

const clearFeedbackLogic = createLogic<State>({
  type: CLEAR_FEEDBACK,
  validate: ({ getState, action }, allow, reject) => {
    const { message, visible } = getState().feedback

    if (message === '' && !visible) {
      reject(rejectClearFeedback())
    }

    allow(action)
  },
})

export default [clearFeedbackLogic]
