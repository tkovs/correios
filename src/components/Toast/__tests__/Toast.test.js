import React from 'react'
import { render, act } from '@testing-library/react-native'
import { createMockStore } from 'redux-logic-test'
import rootReducer from '../../../store/reducers'

import { addFeedback, clearFeedback } from '../../../store/actions/feedback'
import Toast from '..'
import { getComponentWithRedux } from '../../../utils/jest'

jest.useFakeTimers()

describe('Toast component', () => {
  const mockMessage = 'mockMessage'
  const toastDuration = 3000

  it('should react correctly to feedback changes', () => {
    const store = createMockStore({ reducer: rootReducer })
    const { baseElement } = render(getComponentWithRedux(store, <Toast />))

    // Should not be visible initially
    expect(baseElement).toMatchSnapshot()

    // Should be visible with a feedback added
    act(() => {
      store.dispatch(addFeedback(mockMessage))
    })
    expect(store.actions.length).toEqual(1)
    expect(baseElement).toMatchSnapshot()

    // Should not be visible after the feedback was cleaned
    act(() => {
      jest.advanceTimersByTime(toastDuration)
    })
    expect(store.actions.length).toEqual(2)
    expect(baseElement).toMatchSnapshot()

    // Check the store actions
    expect(store.actions[0]).toEqual(addFeedback(mockMessage))
    expect(store.actions[1]).toEqual(clearFeedback())
  })
})
