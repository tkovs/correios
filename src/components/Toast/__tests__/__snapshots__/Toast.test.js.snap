// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast component should be invisible and clear the feedback by clicking the Ok! button 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;

exports[`Toast component should be invisible and clear the feedback by clicking the Ok! button 2`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <SafeAreaView
      pointerEvents="box-none"
      style={
        Object {
          "bottom": 0,
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <View
        accessibilityLiveRegion="polite"
        pointerEvents="box-none"
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#000000",
            "borderRadius": 4,
            "elevation": 6,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "margin": 8,
            "opacity": 0,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 5,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 6,
            "transform": Array [
              Object {
                "scale": 0.9,
              },
            ],
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "flex": 1,
                  "flexWrap": "wrap",
                  "marginLeft": 16,
                  "marginVertical": 14,
                },
                Object {
                  "color": "#ffffff",
                  "marginRight": 0,
                },
              ],
            ]
          }
        >
          mockMessage
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "transparent",
              "borderColor": "transparent",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0,
              "elevation": 0,
              "marginHorizontal": 8,
              "marginVertical": 6,
              "minWidth": "auto",
            }
          }
        >
          <TouchableHighlight
            accessibilityComponentType="button"
            accessibilityRole="button"
            accessibilityStates={Array []}
            accessibilityTraits="button"
            activeOpacity={0.85}
            delayPressIn={0}
            delayPressOut={100}
            disabled={false}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 4,
                },
              ]
            }
            underlayColor="rgba(3, 218, 196, 0.16)"
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "letterSpacing": 1,
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      Object {
                        "marginHorizontal": 8,
                      },
                      Object {
                        "textTransform": "uppercase",
                      },
                      Object {
                        "color": "#03dac4",
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Ok!
              </Text>
            </View>
          </TouchableHighlight>
        </View>
      </View>
    </SafeAreaView>
  </View>
</View>
`;

exports[`Toast component should be invisible and clear the feedback by clicking the Ok! button 3`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;

exports[`Toast component should react correctly to feedback changes 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;

exports[`Toast component should react correctly to feedback changes 2`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <SafeAreaView
      pointerEvents="box-none"
      style={
        Object {
          "bottom": 0,
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <View
        accessibilityLiveRegion="polite"
        pointerEvents="box-none"
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#000000",
            "borderRadius": 4,
            "elevation": 6,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "margin": 8,
            "opacity": 0,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 5,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 6,
            "transform": Array [
              Object {
                "scale": 0.9,
              },
            ],
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "flex": 1,
                  "flexWrap": "wrap",
                  "marginLeft": 16,
                  "marginVertical": 14,
                },
                Object {
                  "color": "#ffffff",
                  "marginRight": 0,
                },
              ],
            ]
          }
        >
          mockMessage
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "transparent",
              "borderColor": "transparent",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0,
              "elevation": 0,
              "marginHorizontal": 8,
              "marginVertical": 6,
              "minWidth": "auto",
            }
          }
        >
          <TouchableHighlight
            accessibilityComponentType="button"
            accessibilityRole="button"
            accessibilityStates={Array []}
            accessibilityTraits="button"
            activeOpacity={0.85}
            delayPressIn={0}
            delayPressOut={100}
            disabled={false}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 4,
                },
              ]
            }
            underlayColor="rgba(3, 218, 196, 0.16)"
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "letterSpacing": 1,
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      Object {
                        "marginHorizontal": 8,
                      },
                      Object {
                        "textTransform": "uppercase",
                      },
                      Object {
                        "color": "#03dac4",
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Ok!
              </Text>
            </View>
          </TouchableHighlight>
        </View>
      </View>
    </SafeAreaView>
  </View>
</View>
`;

exports[`Toast component should react correctly to feedback changes 3`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;
